src/plugins/kdb/ldap/libkdb_ldap/ldap_principal2.c

```diff
@@ -1375,64 +1482,207 @@ krb5_read_tkt_policy(krb5_context context, krb5_ldap_context *ldap_context,
     return st;
 }
 
-krb5_error_code
-krb5_decode_krbsecretkey(krb5_context context, krb5_db_entry *entries,
-                         struct berval **bvalues,
-                         krb5_tl_data *userinfo_tl_data, krb5_kvno *mkvno)
+struct key_entry {
+    krb5_int16 n_key_data;
+    krb5_key_data *key_data;
+};
+
+/*
+ * Decode keys from ldap search results.
+ *
+ * Arguments:
+ *  - bvalues
+ *      The ldap search results containing the key data.
+ *  - mkvno
+ *      The master kvno that the keys were encrypted with.
+ *  - keys_out
+ *      The decoded keys in a key_entry struct. Must be freed.
+ *  - n_keys_out
+ *      The number of key_entry's in keys_out.
+ *  - total_keys_out
+ *      An optional argument that if given will be set to the total number of
+ *      keys found throughout all the entries: sum(keys_out.n_key_data)
+ *      May be NULL.
+ */
+static krb5_error_code
+decode_keys(struct berval **bvalues, krb5_kvno *mkvno,
+            struct key_entry **keys_out, unsigned int *n_keys_out,
+            /* NULL */ unsigned int *total_keys_out)
 {
-    char                        *user=NULL;
-    int                         i=0, j=0, noofkeys=0;
-    krb5_key_data               *key_data=NULL, *tmp;
-    krb5_error_code             st=0;
+    krb5_error_code err = 0;
+    unsigned int n_keys, i, ki, total_keys;
+    struct key_entry *p_keys = NULL;
+
+    *keys_out = NULL;
+    *n_keys_out = 0;
+    if (total_keys_out)
+        *total_keys_out = 0;
 
-    if ((st=krb5_unparse_name(context, entries->princ, &user)) != 0)
+    /*
+     * Get the number of keys. Previously, similar implementations used realloc
+     * instead of pre-counting but I think this will have better performance.
+     */
+    for (n_keys = 0, i = 0; bvalues[i] != NULL; i++)
```

Comment:
Performance is irrelevant here; the cost of LDAP queries will dominate any time we spend on memory bookkeeping in this function.  If you want to pre-count for simplicity that's fine, but performance is not a meaningful justification for it.